File: 3sum/quadratic.py
# lol "3sum"


from typing import List


# leetcode.com boilerplate
class Solution:
    def threeSum(self, nums: List[int]) -> List[List[int]]:
        return list(null_triples(nums))


def print(*args, **kwargs):
    pass


def null_triples(numbers):
    """Yield all distinct triples from `numbers` whose sum is zero."""
    lookup = lookup_table(numbers)
    yield from three_sum(numbers, lookup, 0)


def three_sum(numbers, lookup, k):
    solutions = set()  # set of sets of three numbers

    for i in range(len(numbers)):
        print(f'looking at index {i}, which is {numbers[i]}')
        print(f'That means the sum to look for is {k - numbers[i]}')
        for (j, m) in two_sum_index(numbers, lookup, k - numbers[i]):
            print(f'two_sum_index({k - numbers[i]}) -> {(j, m)}')
            if i in (j, m):
                continue

            values    = [numbers[i], numbers[j], numbers[m]]
            candidate = frozenset(values)
            if candidate not in solutions:
                solutions.add(candidate)
                yield values


def lookup_table(numbers):
    table = {}
    for i, number in enumerate(numbers):
        table.setdefault(number, set()).add(i)

    return table


def two_sum_index(numbers, lookup, k):
    solutions = set()  # set of sets of two numbers
    print(f'two_sum_index({k})')

    for i in range(len(numbers)):
        indices = lookup.get(k - numbers[i])
        print(f'lookup.get({k - numbers[i]}) -> {indices}')
        if indices is None:
            continue

        for j in indices:
            if j == i:
                continue

            values    = [numbers[i], numbers[j]]
            candidate = frozenset(values)
            if candidate not in solutions:
                solutions.add(candidate)
                yield [i, j]

File: 3sum/quadratic.cpp
#include <algorithm>
#include <set>
#include <unordered_map>
#include <vector>

std::vector<std::vector<int>> threeSum(const std::vector<int>& numbers);

// leetcode.com boilerplate
class Solution {
public:
    std::vector<std::vector<int>> threeSum(std::vector<int>& nums) {
        return ::threeSum(nums);
    }
};

std::unordered_multimap<int, int> tabulate(const std::vector<int>& numbers) {
    std::unordered_multimap<int, int> result;

    for (int i = 0; i < int(numbers.size()); ++i) {
        result.emplace(numbers[i], i);
    }

    return result;
}

std::vector<std::vector<int>> threeSum(const std::vector<int>& numbers) {
    const std::unordered_multimap<int, int> indicesByValue = tabulate(numbers);

    std::set<std::vector<int>> solutions;

    const int n = numbers.size();
    for (int i = 0; i < n; ++i) {
        for (int j = i + 1; j < n; ++j) {
            const auto [begin, end] = indicesByValue.equal_range(-numbers[i] - numbers[j]);
            if (begin == indicesByValue.end()) {
                continue;
            }

            for (auto iter = begin; iter != end; ++iter) {
                const int k = iter->second;
                if (k == i || k == j) {
                    continue;
                }

                std::vector<int> candidate{numbers[i],
                                           numbers[j],
                                           numbers[k]};
                std::sort(candidate.begin(), candidate.end());
                solutions.insert(candidate);
            }
        }
    }

    return std::vector<std::vector<int>>(solutions.begin(), solutions.end());
}